File: src/core/project-overview.ts
/**
 * Per-project summaries for the all-projects lens.
 *
 * The Git, GitHub, and Vercel pages each read the daemon's *selected*
 * repository, which is global state — so "show me all projects" cannot be
 * served by asking those pages nicely. Everything here resolves a project from
 * its own path instead and never touches `selectedGitRepository`, which is what
 * makes an overview possible without the act of looking changing what the rest
 * of the app is pointed at.
 *
 * One project failing is normal (no remote, no Vercel link, a token that does
 * not cover it), so a failure is carried on that project's own row rather than
 * failing the whole request.
 */

import type { ConfigStore } from "./config-store.js";
import { GitManager } from "./git-manager.js";
import { optionalGitHubContext } from "./github-context.js";
import { optionalProviderContext } from "./providers/registry.js";
import type { ProviderDeploymentState } from "./providers/deploy-provider.js";
import type { GitRepositoryDefinition } from "./types.js";

export type OverviewDomain = "git" | "github" | "vercel";

export interface ProjectGitSummary {
  branch: string;
  /** Files with any working-tree or index change, including untracked. */
  dirty: number;
  ahead: number;
  behind: number;
  upstream?: string;
}

export interface ProjectGitHubSummary {
  owner: string;
  repo: string;
  openPullRequests: number;
  /** CI on the default branch's head, when it could be read. */
  checks?: "success" | "failure" | "pending";
}

export interface ProjectVercelSummary {
  projectId: string;
  projectName: string;
  /** Absent when the project has never deployed. */
  state?: ProviderDeploymentState;
  url?: string;
  createdAt?: number;
}

export interface ProjectOverviewEntry {
  name: string;
  path: string;
  /** The worktree actually read, which may differ from `path`. */
  cwd: string;
  git?: ProjectGitSummary;
  github?: ProjectGitHubSummary;
  vercel?: ProjectVercelSummary;
  /** Why this project has no summary. Rendered on the card, not thrown. */
  error?: string;
}

/**
 * How many projects are summarized concurrently.
 *
 * Each GitHub row costs two API calls and each Vercel row up to three, so an
 * unbounded fan-out over a large workspace would burst straight into a rate
 * limit. Six keeps a realistic workspace responsive without doing that.
 */
const CONCURRENCY = 6;

export async function buildProjectOverview(
  configStore: ConfigStore,
  domain: OverviewDomain,
): Promise<ProjectOverviewEntry[]> {
  const config = await configStore.load();
  return mapWithLimit(config.gitRepositories, CONCURRENCY, (repository) =>
    summarize(configStore, repository, domain),
  );
}

async function summarize(
  configStore: ConfigStore,
  repository: GitRepositoryDefinition,
  domain: OverviewDomain,
): Promise<ProjectOverviewEntry> {
  const cwd = repository.activeWorktreePath ?? repository.path;
  const base = { name: repository.name, path: repository.path, cwd };
  try {
    switch (domain) {
      case "git":
        return { ...base, git: await gitOverview(cwd) };
      case "github":
        return { ...base, github: await githubOverview(configStore, cwd) };
      case "vercel":
        return { ...base, vercel: await vercelOverview(configStore, cwd) };
    }
  } catch (error) {
    return { ...base, error: error instanceof Error ? error.message : String(error) };
  }
}

async function gitOverview(cwd: string): Promise<ProjectGitSummary> {
  const status = await new GitManager(cwd).status();
  return {
    branch: status.branch,
    dirty: status.files.length,
    ahead: status.ahead,
    behind: status.behind,
    upstream: status.upstream,
  };
}

async function githubOverview(configStore: ConfigStore, cwd: string): Promise<ProjectGitHubSummary> {
  const context = await optionalGitHubContext(configStore, cwd);
  if (!context) throw new Error("No GitHub repository resolves for this project.");

  const prs = await context.manager.listPRs("open");
  // CI is a nice-to-have on a card: a repo with no Actions, or a token without
  // the checks scope, should still show its PR count rather than an error.
  const checks = await context.manager
    .repoInfo()
    .then((info) => context.manager.getCommitChecks(info.default_branch ?? "HEAD"))
    .then((result) => normalizeChecks(result.state))
    .catch(() => undefined);

  return {
    owner: context.owner,
    repo: context.repo,
    openPullRequests: prs.length,
    checks,
  };
}

/** The card only distinguishes three outcomes; anything else reads as unknown. */
function normalizeChecks(state: string): ProjectGitHubSummary["checks"] {
  if (state === "success") return "success";
  if (state === "failure") return "failure";
  if (state === "pending" || state === "in_progress" || state === "queued") return "pending";
  return undefined;
}

async function vercelOverview(configStore: ConfigStore, cwd: string): Promise<ProjectVercelSummary> {
  const context = await optionalProviderContext("vercel", configStore, cwd);
  if (!context?.project) throw new Error("No Vercel project is linked to this project.");

  // Production only, and only the latest: the card answers "is what users see
  // healthy", not "what has been built lately".
  const [latest] = await context.provider.listDeployments({
    projectId: context.project.id,
    target: "production",
    limit: 1,
  });

  return {
    projectId: context.project.id,
    projectName: context.project.name,
    state: latest?.state,
    url: latest?.url ?? undefined,
    createdAt: latest?.createdAt,
  };
}

/**
 * `Promise.all` over a bounded worker pool, preserving input order. Workers
 * pull from a shared cursor, so a slow project delays only itself rather than
 * holding up a whole batch the way chunked `Promise.all` rounds would.
 */
async function mapWithLimit<T, R>(
  items: readonly T[],
  limit: number,
  map: (item: T) => Promise<R>,
): Promise<R[]> {
  const results = new Array<R>(items.length);
  let cursor = 0;

  async function worker(): Promise<void> {
    while (cursor < items.length) {
      const index = cursor++;
      results[index] = await map(items[index] as T);
    }
  }

  await Promise.all(
    Array.from({ length: Math.min(limit, items.length) }, () => worker()),
  );
  return results;
}

File: crates/nomoreide-core/src/project_overview.rs
//! Per-project summaries for the all-projects lens.
//!
//! The Rust half of `src/core/project-overview.ts`. The Git, GitHub and Vercel
//! pages each read the daemon's *selected* repository, which is global state —
//! so "show me all projects" cannot be served by asking those pages nicely.
//! Everything here resolves a project from its own path instead and never
//! touches the selection, which is what lets an overview exist without the act
//! of looking changing what the rest of the app is pointed at.
//!
//! **One project failing is normal**: no remote, no Vercel link, a token that
//! does not cover it. A failure therefore rides on that project's own row and
//! the request still succeeds, so one unreachable project cannot blank a page
//! that is mostly about the others.

use crate::config::{ConfigStore, GitRepoDef};
use crate::git_manager::GitManager;
use serde_json::{Map, Value};

/// The three lenses the page offers. Anything else is a 404 at the route.
pub const OVERVIEW_DOMAINS: [&str; 3] = ["git", "github", "vercel"];

pub async fn build_project_overview(
    store: &ConfigStore,
    domain: &str,
) -> Result<Vec<Value>, String> {
    let config = store.load().await.map_err(|error| error.to_string())?;
    let mut rows = Vec::with_capacity(config.git_repositories.len());
    // Sequential where the reference runs six at a time. The bound is there to
    // keep a large workspace from bursting into a rate limit, not to change the
    // answer: results keep input order either way.
    for repository in &config.git_repositories {
        rows.push(summarize(store, repository, domain).await);
    }
    Ok(rows)
}

async fn summarize(store: &ConfigStore, repository: &GitRepoDef, domain: &str) -> Value {
    // The worktree actually read, which may not be the repository's own path.
    let cwd = repository
        .active_worktree_path
        .clone()
        .unwrap_or_else(|| repository.path.clone());

    let mut row = Map::new();
    row.insert("name".into(), Value::String(repository.name.clone()));
    row.insert("path".into(), Value::String(repository.path.clone()));
    row.insert("cwd".into(), Value::String(cwd.clone()));

    let summary = match domain {
        "git" => git_overview(&cwd).await.map(|value| ("git", value)),
        "github" => github_overview(store, &cwd).await.map(|v| ("github", v)),
        "vercel" => vercel_overview(store, &cwd).await.map(|v| ("vercel", v)),
        _ => Err(format!("Unknown overview domain: {domain}")),
    };
    match summary {
        Ok((key, value)) => {
            row.insert(key.into(), value);
        }
        Err(message) => {
            row.insert("error".into(), Value::String(message));
        }
    }
    Value::Object(row)
}

async fn git_overview(cwd: &str) -> Result<Value, String> {
    let status = GitManager::status(cwd).await.map_err(|e| e.to_string())?;
    let mut summary = Map::new();
    summary.insert("branch".into(), Value::String(status.branch));
    // Any index or working-tree change, untracked included: the card counts
    // "things you have not committed", not "files git will commit".
    summary.insert("dirty".into(), Value::from(status.files.len()));
    summary.insert("ahead".into(), Value::from(status.ahead));
    summary.insert("behind".into(), Value::from(status.behind));
    if let Some(upstream) = status.upstream {
        summary.insert("upstream".into(), Value::String(upstream));
    }
    Ok(Value::Object(summary))
}

async fn github_overview(store: &ConfigStore, cwd: &str) -> Result<Value, String> {
    let Some(context) = crate::github_context::optional_github_context(store, cwd).await else {
        return Err("No GitHub repository resolves for this project.".into());
    };
    let prs = context
        .manager
        .list_prs("open", "1")
        .await
        .map_err(|error| error.to_string())?;

    // CI is a nice-to-have on a card: a repository with no Actions, or a token
    // without the checks scope, should still show its pull-request count rather
    // than turning the whole row into an error.
    let checks = match context.manager.repo_info().await {
        Ok(info) => {
            let branch = info
                .get("default_branch")
                .and_then(Value::as_str)
                .unwrap_or("HEAD")
                .to_string();
            context
                .manager
                .commit_checks(Some(&branch))
                .await
                .ok()
                .and_then(|status| normalize_checks(&status.state))
        }
        Err(_) => None,
    };

    let mut summary = Map::new();
    summary.insert("owner".into(), Value::String(context.owner));
    summary.insert("repo".into(), Value::String(context.repo));
    summary.insert("openPullRequests".into(), Value::from(prs.len()));
    if let Some(checks) = checks {
        summary.insert("checks".into(), Value::String(checks.into()));
    }
    Ok(Value::Object(summary))
}

/// The card distinguishes three outcomes; anything else reads as unknown and is
/// left off rather than guessed at.
fn normalize_checks(state: &str) -> Option<&'static str> {
    match state {
        "success" => Some("success"),
        "failure" => Some("failure"),
        "pending" | "in_progress" | "queued" => Some("pending"),
        _ => None,
    }
}

async fn vercel_overview(store: &ConfigStore, cwd: &str) -> Result<Value, String> {
    let config = store.load().await.map_err(|error| error.to_string())?;
    // A provider that will not connect and a repository with no linked project
    // are the same answer here: there is nothing to put on the card.
    let context =
        crate::providers::registry::require_provider_context("vercel", store, &config, cwd)
            .await
            .map_err(|_| "No Vercel project is linked to this project.".to_string())?;
    let Some(project) = context.project.as_ref() else {
        return Err("No Vercel project is linked to this project.".into());
    };
    let Some(identifier) = project.id.as_ref().and_then(Value::as_str) else {
        return Err("No Vercel project is linked to this project.".into());
    };

    // Production only, and only the latest: the card answers "is what users see
    // healthy", not "what has been built lately".
    let latest = context
        .client
        .list_deployments(identifier, Some("production"), 1)
        .await
        .map_err(|error| error.to_string())?
        .into_iter()
        .next();

    let mut summary = Map::new();
    summary.insert("projectId".into(), Value::String(identifier.to_string()));
    summary.insert(
        "projectName".into(),
        project.name.clone().unwrap_or(Value::Null),
    );
    if let Some(deployment) = latest {
        let raw = serde_json::to_value(&deployment).unwrap_or(Value::Null);
        for key in ["state", "url", "createdAt"] {
            if let Some(value) = raw.get(key).filter(|value| !value.is_null()) {
                summary.insert(key.into(), value.clone());
            }
        }
    }
    Ok(Value::Object(summary))
}
